modules/objdetect/src/qrcode.cpp

```diff
@@ -3816,33 +3835,469 @@ bool QRDetectMulti::computeTransformationPoints(const size_t cur_ind)
+        Point2f toOther = getDirectionTo(other.typePattern);
+        Point2f toThis = other.getDirectionTo(typePattern);
+        const float cosAngle = getCosAngle(toOther, toThis);
+        if (cosAngle < 0.f && (CV_PI - acos(cosAngle)) / 2.f < maxRotateDiff) {
+            const float angleCenter = max(acos(getCosAngle(toOther, other.center - center)), acos(getCosAngle(toThis, center - other.center)));
+            if (angleCenter < maxRotateDiff)
+                return true;
+            }
         return false;
     }
-    vector< vector< Point2f > > pnts2f = qrdet.getTransformationPoints();
-    vector<Point2f> trans_points;
-    for(size_t i = 0; i < pnts2f.size(); i++)
-        for(size_t j = 0; j < pnts2f[i].size(); j++)
-            trans_points.push_back(pnts2f[i][j]);
 
-    updatePointsResult(points, trans_points);
+    static float getCosAngle(Point2f vec1, Point2f vec2) {
+        float cosAngle = vec1.dot(vec2) / (sqrt(normL2Sqr<float>(vec1)) * sqrt(normL2Sqr<float>(vec2)));
+        if (cosAngle < -1.f)
+            cosAngle = -1.f;
+        else if (cosAngle > 1.f)
+            cosAngle = 1.f;
+        return cosAngle;
+    }
 
-    return true;
+    pair<int, Point2f> getQRCorner() const {
+        if (typePattern == TypePattern::CENTER) {
+            int id = (innerCornerId + 2) % 4;
+            return std::make_pair(id, points[id]);
+        }
+        else if (typePattern != TypePattern::NONE) {
+            int id = (innerCornerId + 2) % 4;
+            return std::make_pair(id, points[id]);
+        }
+        return std::make_pair(-1, Point2f());
+    }
+
+    pair<int, Point2f> getCornerForIntersection() const {
+        if (typePattern == TypePattern::RIGHT) {
+            int id = (innerCornerId + 3) % 4;
+            return std::make_pair(id, points[id]);
+        }
+        else if (typePattern == TypePattern::BOTTOM) {
+            int id = (innerCornerId + 1) % 4;
+            return std::make_pair(id, points[id]);
+        }
+        return std::make_pair(-1, Point2f());
+    }
+
+    Point2f getTimingStart(TypePattern direction) const {
+        Point2f start = points[innerCornerId]*(6.5f/7.f);
+        if (typePattern == TypePattern::CENTER && direction == TypePattern::RIGHT) {
+            start += points[(innerCornerId + 3) % 4]*(0.5f/7.f);
+        }
+        else if (typePattern == TypePattern::CENTER && direction == TypePattern::BOTTOM) {
+            start += points[(innerCornerId + 1) % 4]*(0.5f/7.f);
+        }
+        else if (typePattern == TypePattern::RIGHT && direction == TypePattern::CENTER) {
+            start += points[(innerCornerId + 1) % 4]*(0.5f/7.f);
+        }
+        else if (typePattern == TypePattern::BOTTOM && direction == TypePattern::CENTER) {
+            start += points[(innerCornerId + 3) % 4]*(0.5f/7.f);
+        }
+        return start + getDirectionTo(direction)/(7.f*2.f);
+    }
+
+    // return (total white+black pins, total first color, penaltyPoints)
```

Comment:
What "total first color" means?

Reply:
renaming to `return total white+black modules in timing pattern, total white modules, penaltyPoints`